apps/pipelines/tests/test_code_node.py

```diff
@@ -0,0 +1,88 @@
+import json
+from unittest import mock
+
+import pytest
+
+from apps.pipelines.exceptions import PipelineNodeBuildError, PipelineNodeRunError
+from apps.pipelines.nodes.base import PipelineState
+from apps.pipelines.tests.utils import (
+    code_node,
+    create_runnable,
+    end_node,
+    start_node,
+)
+from apps.utils.factories.pipelines import PipelineFactory
+from apps.utils.pytest import django_db_with_data
+
+
+@pytest.fixture()
+def pipeline():
+    return PipelineFactory()
+
+
+EXTRA_FUNCTION = """
+def other(foo):
+    return f"other {foo}"
+
+return other(input)
+"""
+
+IMPORTS = """
+import json
+import datetime
+import re
+import time
+return json.loads(input)
+"""
+
+
+@django_db_with_data(available_apps=("apps.service_providers",))
+@mock.patch("apps.pipelines.nodes.base.PipelineNode.logger", mock.Mock())
+@pytest.mark.parametrize(
+    ("code", "input", "output"),
+    [
+        ("return f'Hello, {input}!'", "World", "Hello, World!"),
+        ("", "foo", "foo"),  # No code just returns the input
+        (EXTRA_FUNCTION, "blah", "other blah"),  # Calling a separate function is possible
+        ("'foo'", "", "None"),  # No return value will return "None"
```

Comment:
I'm wondering if we should instead raise an exception in this case, as it was probably not intentional to not return a string.

Reply:
I think it makes sense to raise. How easy would it be to detect this when saving?